Add support for client-side routing in both deployment systems

samderanova opened the pull request:
Closes #17 by creating an exception handler to redirect all requests to the container that result in a status code of 404 to `index.html` (the frontend application). This way, the frontend can handle the routing.

For deployment on Vercel, a [rewrite rule for SPA Fallback](https://vercel.com/docs/projects/project-configuration#legacy-spa-fallback) is added to `vercel.json`.

## Testing
1. Install `react-router-dom` in the frontend application.
2. Set up the default route `/` to the default component (e.g. `<App />`).
3. Create a new route and a new component associated with that route. For example, you can create the `/test` route that points to a new component `TestPage` in `TestPage.jsx`.
4. Build the Docker container locally with `docker build -t starter-pack .`
5. Run the container while exposing port 8000, the port that the Dockerfile runs the FastAPI application on: `docker run -p 8000:8000 starter-pack`.
6. Navigate to `http://localhost:8000/[new-route]` and observe the component associated with that route being rendered onto the screen.

samderanova committed:
Add support for client-side routing

- Create exception handler to redirect all requests to the container
  that result in a status code of 404 to index.html (the frontend)

samderanova committed:
Use HTTP_404_NOT_FOUND rather than constant 404

taesungh commented on backend/src/main.py:
```diff
@@ -36,4 +36,8 @@ async def root() -> FileResponse:
 app.mount("/", StaticFiles(directory=PUBLIC_DIRECTORY), name="public")
 
+
 # TODO: add support for client-side routing
```
Let's replace the comment and make the handler opt-in. We can instruct users to uncomment the following if using SPA client-side routing, e.g. React Router.

samderanova replied:
Oh what I thought I removed this comment. Noted, I'll replace it with that.

samderanova replied:
Update: since a significant number of teams will most likely be using `react-router` for their projects, we should leave this opt-out (i.e leaving it uncommented).

taesungh commented on backend/src/main.py:
```diff
@@ -36,4 +36,8 @@ async def root() -> FileResponse:
+
 # TODO: add support for client-side routing
+@app.exception_handler(status.HTTP_404_NOT_FOUND)
+async def not_found(req: Request, exc: HTTPException) -> FileResponse:
+    return FileResponse(PUBLIC_DIRECTORY / "index.html")
```
Chore: let's include a docstring explaining that the SPA bundle is sent for the client/browser to resolve.

taesungh commented on backend/src/main.py:
```diff
@@ -12,12 +12,12 @@
 from fastapi.responses import FileResponse
 from fastapi.staticfiles import StaticFiles
 
-import api
-
```
Let's keep the first-party import where it was.
If you have isort running locally, need to specify option for src paths.

taesungh commented on backend/src/main.py:
```diff
@@ -36,4 +36,8 @@ async def root() -> FileResponse:
 
+
 # TODO: add support for client-side routing
+@app.exception_handler(status.HTTP_404_NOT_FOUND)
```
Praise: thanks for using the named status value.

samderanova committed:
Add docstring for not_found catch all

samderanova committed:
Update Vercel Rewrite

- Now that the exception handler has been setup, add an additional
  rewrite in vercel.json to rewrite all requests not directed to /api/
  to /index.html

taesungh committed:
Update docstring and notes in `DEPLOYMENT.md`

taesungh reviewed (approved):
Thanks for adding and finding a solution that works with FastAPI.
